doto-frontend/src/components/pages/Calendar/TaskScheduler.js

```diff
@@ -0,0 +1,73 @@
+const MILLISECONDS_PER_MINUTE = 60000;
+
+/**
+ * Takes in a single new unscheduled task as well as a list of existing scheduled tasks and returns
+ * a newly-scheduled, full list of tasks. The existing tasks may have their schedules modified.
+ *
+ * NOTE: This function assumes existingTasks are ordered chronologically by startDate, i.e. have been
+ * processed using this function.
+ *
+ * @param {object} newTasks The new tasks to schedule, with no start/end datetimes present
+ * @param {Array<object>} existingTasks The existing scheduled tasks, with start/end datetimes present
+ * @param {Date} currDate The datetime after which the task should be scheduled (usually the current datetime)
+ * @returns A chronologically ordered (based on startDate) array of all tasks scheduled with start/end
+ * datetimes - essentially existingTasks + newTasks
+ */
+const addTaskToSchedule = (newTask, existingTasks, currDate) => {
+    // TODO: Take into account any possible gap between datetime and startDate of the first task in oldTasks
+    // TODO: Take into account priority of tasks
+    // TODO: Take into account location of tasks, add time gaps to allow for travel
+
+    const competingTasks = [];
+    const oldTasks = [];
+
+    // Separate existing tasks with a startDate > current datetime from older tasks (which won't be considered)
+    existingTasks.forEach(task => {
+        task.startDate > currDate ? competingTasks.push(task) : oldTasks.push(task);
+    });
+
+    // If the endDate of the latest oldTask is after currDate, then the earliest any new task can be scheduled is oldTask.endDate
```

Comment:
is there any reason why the start date of the new task can't be earlier than the end date of the latest old task?

Reply:
The current time is after the start date of the latest old task, meaning the latest old task is currently being done

Reply:
I guess we don't want users to add tasks that are overlapping with each other